test/client/doc.js

```diff
@@ -214,6 +214,45 @@ describe('Doc', function() {
       });
     });
 
+    it('remote ops are transformed by ops submitted in `before op` event handlers', function(done) {
+      var doc = this.doc;
+      var doc2 = this.doc2;
+      var doc3 = this.doc3;
+      doc2.submitOp({p: ['list'], oi: []}, function() {
+        doc.fetch(function() {
+          var opFromBeforeOpEvent = 1;
+          doc.on('before op', function(op, source) {
+            // console.log('\nbefore op @ doc', op, source ? 'local' : 'remote')
+            if (source) {
+              return;
+            }
+            if (opFromBeforeOpEvent <= 0) {
+              return;
+            }
+            opFromBeforeOpEvent--;
+            doc.submitOp({p: ['list', 0], li: 2}, {source: true});
+          });
```

Comment:
You can make this a bit more readable by getting rid of the counter:

```suggestion
          function beforeOpHandler(op, source) {
            if (source) return;
            doc.off('before op', beforeOpHandler);
            doc.submitOp({p: ['list', 0], li: 2}, {source: true});
          }
          doc.on('before op', beforeOpHandler);
```

Same thing below, with `'op'`.

Reply:
done.